Updates dependencies to get us pretty close to current

tnthornton opened the pull request:
### Description of your changes
It's been a long time since dependencies were updated in this repo. Some of the consequences of that are:
- [x] The crossplane/crossplane dependency hasn't been updated since 1.6 resulting in issues such as:
     * #329 
- [x] Any security issues that may have cropped over the last "months" time would remain unresolved.
- [x] The general PR, as seen here, is much larger changeset than we'd tend to like.

The overarching goal of this PR is to get us roughly caught up on the crossplane/crossplane dependency which results in a cascade affect of the other dependencies needing to be updated.

In addition:
* The minimum Go version is being updated to 1.20 (which has affects on the GolangCI Lint version).
* The build submodule is also updated.

Fixes #329

I have:

- [X] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested
1. Basic validation: `make build test reviewable`
2. More extensive tests are in progress:
- [X] xpkg APIs, ensuring above linked issues are resolved.
- [x] xpls still works the way it did prior to this update


#### Testing xpls
1. `make build`
3. move up into terminal path.
4. Check version
```
up --version
v0.18.0-rc.0.11.g6bceab2
```
See below screenshots, prior functionality appears to continue to work
![Screen Shot 2023-06-23 at 5 23 49 PM](https://github.com/upbound/up/assets/2375126/b51439d0-f9ea-45be-8e45-1506c6912905)
![Screen Shot 2023-06-23 at 5 24 41 PM](https://github.com/upbound/up/assets/2375126/9d33ba81-6606-41b7-a481-e6cef24236bc)
![Screen Shot 2023-06-23 at 5 23 35 PM](https://github.com/upbound/up/assets/2375126/cb4c785f-688c-4c90-8725-eafe863653a5)
![Screen Shot 2023-06-23 at 5 23 09 PM](https://github.com/upbound/up/assets/2375126/69489bda-b6f6-4465-a7a3-f42396793610)
![Screen Shot 2023-06-23 at 5 16 52 PM](https://github.com/upbound/up/assets/2375126/ae3d2596-32fe-47a4-8f14-0534725f4c8c)

#### Testing xpkg commands
a. Checking if #329 is resolved

1. Followed repro steps in #329 and ended up with an `ezgi-platform-ref-gcp.xpkg` file.
2. Using up, ran the following command to extract the package `up xpkg xp-extract --from-xpkg ezgi-platform-ref-gcp.xpkg`
3. Used gzip to uncompress the resulting out.gz: `gzip --uncompress out.gz`
4. Opened the `out` file, then searched for `status.atProvider.id` and found:
```
    - fromFieldPath: status.atProvider.id
      policy:
        fromFieldPath: Required
      toFieldPath: status.gke.project
      transforms:
      - string:
          regexp:
            group: 1
            match: projects\/(.+)\/serviceAccounts\/.*
          type: Regexp
        type: string
      type: ToCompositeFieldPath
```

b. Verified that I can push xpkg to registry
```
up xpkg push xpkg.upbound.io/taylor/tt-test:v0.0.1 -f ezgi-platform-ref-gcp.xpkg
xpkg pushed to xpkg.upbound.io/taylor/tt-test:v0.0.1
```

![Screen Shot 2023-06-23 at 6 02 27 PM](https://github.com/upbound/up/assets/2375126/036403b7-e0ff-424f-ada1-3329bdb39ec9)

c. Verified that I can pull that xpkg into a control plane
```
[kubedebug] / # cat c.yaml
apiVersion: pkg.crossplane.io/v1
kind: Configuration
metadata:
  name: platform-ref-gcp
spec:
  package: xpkg.upbound.io/taylor/tt-test:v0.0.1


[kubedebug] / # kubectl get providers
NAME                                     INSTALLED   HEALTHY   PACKAGE                                                          AGE
crossplane-contrib-provider-helm         True        True      xpkg.upbound.io/crossplane-contrib/provider-helm:v0.15.0         5m35s
upbound-provider-aws                     True        Unknown   xpkg.upbound.io/upbound/provider-aws:v0.36.0                     5m52s
upbound-provider-family-gcp              True        True      xpkg.upbound.io/upbound/provider-family-gcp:v0.33.1              5m11s
upbound-provider-gcp-cloudplatform       True        True      xpkg.upbound.io/upbound/provider-gcp-cloudplatform:v0.33.0       5m32s
upbound-provider-gcp-compute             True        True      xpkg.upbound.io/upbound/provider-gcp-compute:v0.33.0             5m29s
upbound-provider-gcp-container           True        True      xpkg.upbound.io/upbound/provider-gcp-container:v0.33.0           5m23s
upbound-provider-gcp-servicenetworking   True        True      xpkg.upbound.io/upbound/provider-gcp-servicenetworking:v0.33.0   5m19s
upbound-provider-gcp-sql                 True        True      xpkg.upbound.io/upbound/provider-gcp-sql:v0.33.0                 5m15s


[kubedebug] / # kubectl get providerrevisions
NAME                                                  HEALTHY   REVISION   IMAGE                                                            STATE    DEP-FOUND   DEP-INSTALLED   AGE
crossplane-contrib-provider-helm-503c3591121b         True      1          xpkg.upbound.io/crossplane-contrib/provider-helm:v0.15.0         Active                               5m30s
upbound-provider-aws-3f7547b28ca6                               1          xpkg.upbound.io/upbound/provider-aws:v0.36.0                     Active                               5m48s
upbound-provider-family-gcp-b0fb8e3d4a7a              True      1          xpkg.upbound.io/upbound/provider-family-gcp:v0.33.1              Active                               5m5s
upbound-provider-gcp-cloudplatform-78133325933e       True      1          xpkg.upbound.io/upbound/provider-gcp-cloudplatform:v0.33.0       Active   1           1               5m27s
upbound-provider-gcp-compute-308b9eb8c1d3             True      1          xpkg.upbound.io/upbound/provider-gcp-compute:v0.33.0             Active   1           1               5m24s
upbound-provider-gcp-container-fbc4b73cd0e7           True      1          xpkg.upbound.io/upbound/provider-gcp-container:v0.33.0           Active   1           1               5m18s
upbound-provider-gcp-servicenetworking-3b85e9b62814   True      1          xpkg.upbound.io/upbound/provider-gcp-servicenetworking:v0.33.0   Active   1           1               5m14s
upbound-provider-gcp-sql-d88e7407bac4                 True      1          xpkg.upbound.io/upbound/provider-gcp-sql:v0.33.0                 Active   1           1               5m10s
```

AlainRoy commented on cmd/up/main.go:
```diff
@@ -139,7 +139,8 @@ type alpha struct {
 
 func main() {
-	rand.Seed(time.Now().UTC().UnixNano())
+	// We don't need this choice to be cryptographically random.
+	rand.New(rand.NewSource(time.Now().UTC().UnixNano())) // nolint:gosec
```
Just because I'm curious: If we don't need it to be cryptographically strong, why do we need random? What do we use it for?

tnthornton replied:
Funny enough, in searching for the answer to your question it looks like it was being used for some now dead code: https://github.com/upbound/up/commit/c4c3627813860700fdf67cd547f7ae6c52c7567a. I just removed this line to remove the confusion.

AlainRoy commented on go.mod:
```diff
@@ -1,254 +1,241 @@
 	github.com/willabides/kongplete => ./internal/vendor/github.com/WillAbides/kongplete
+)
 
-	// The following replacements should be kept in sync with upstream
-	// Kubernetes version.
-	// xref: https://github.com/kubernetes/kubernetes/blob/4ce5a8954017644c5420bae81d72b09b735c21f0/go.mod#L394
```
Yay! I'm happy to see these replacements removed.

AlainRoy commented on go.mod:
```diff
@@ -1,254 +1,241 @@
-	go.opentelemetry.io/otel/trace => go.opentelemetry.io/otel/trace v0.20.0
-	go.opentelemetry.io/proto/otlp => go.opentelemetry.io/proto/otlp v0.7.0
+replace (
+	// https://github.com/helm/helm/issues/11821
+	github.com/docker/cli => github.com/docker/cli v20.10.19+incompatible
+	github.com/docker/docker => github.com/docker/docker v20.10.19+incompatible
```
Ewww. 
I know it's not your fault and it looks like we have to do this, but still. Eww.

AlainRoy commented:
LGTM! I'll wait to approve until you test, but it looks reasonable. 

My only concern is about what happens for new releases of Crossplane. Will it be a lot of work to update the vendored code? Would it help to document anything about the process you followed to do the update, or is just something that we need to work through?

tnthornton commented:
>My only concern is about what happens for new releases of Crossplane. Will it be a lot of work to update the vendored code? Would it help to document anything about the process you followed to do the update, or is just something that we need to work through?

It's incredibly unclear what the LOE is to update the next time around. Regarding documentation, I don't really know what to document:
* I updated a dependency.
* Fixed compilation errors.
* Copied code over and tore out things that "we didn't need". Still TBD if that plan worked.

AlainRoy commented:
> > My only concern is about what happens for new releases of Crossplane. Will it be a lot of work to update the vendored code? Would it help to document anything about the process you followed to do the update, or is just something that we need to work through?
> 
> It's incredibly unclear what the LOE is to update the next time around. Regarding documentation, I don't really know what to document:
> 
>     * I updated a dependency.
> 
>     * Fixed compilation errors.
> 
>     * Copied code over and tore out things that "we didn't need". Still TBD if that plan worked.

OK, that's cool. No need to do anything to document it.

AlainRoy commented:
Should we do a new release of up in the near future?

tnthornton commented:
> Should we do a new release of up in the near future

That makes sense to me. Maybe sometime next week so that we can at the very least get fixes for #329 out.

jastang reviewed (approved):
Thanks @tnthornton ! Did a few more tests with `xpkg xp-extract` with a local build and looks good 👍🏼 

@phisco did you want this backported?

phisco commented:
@jastang I don't think we support multiple releases, looking at https://github.com/upbound/up/tags it looks like we are always going forward with versions, so nothing to backport to
